### DISCUSSION TITLE
Future x86 emulator

### DISCUSSION THREAD

**Discussion by miguelmazetto** [AndreRH/hangover#134] [General]
I managed to port the new version of hangover to a android termux package, and by doing so i discovered that there is quite alot missing in this version of qemu, or there is something i dont understand here. This is an attempt of installing Firefox, apparently even the arm64 version they distribute on their website contains x86 binaries, so this is what happens when you try to install it:
![winefirefoxsc](https://user-images.githubusercontent.com/33629994/222312215-6cd189e5-644c-45c0-9c5b-54b4ce6bc831.png)
And this is what the terminal says:
![winefirefoxterm](https://user-images.githubusercontent.com/33629994/222312299-90ca7cea-f7fc-451b-9a43-1fe63d60f17a.png)

OBS: no proot/chroot btw
Termux Package: https://github.com/miguelmazetto/termux-packages-wine/tree/master/packages/hangover

**Comment by miguelmazetto** [comment]
Since aparently steam doesnt care about the issue that firefox is complaining, im going ahead to doing a steam deck app for android. But i will be back to contribute to this repo about x86_64 emulation.
![image](https://user-images.githubusercontent.com/33629994/222552468-02aa0067-5d1d-41b2-b21b-343e61f85c87.png)

**Comment by AndreRH** [comment]
Hi,
 - regarding SSE2, this isn't about Qemu, but about ARM64 Wine not yet advertising the support. ~~Will be fixed with the next push/release~~ Fixed now. (though the Firefox Installer still won't run). 
 - regarding the x86_64 emulation, WoW64 actually means "Windows32 on Windows64", so that's not the way for 64-bit, but there's already a plan for upstream Wine to support that by using a mechanism similar to what Windows does. It might be possible to make a "W64oW64.dll", but I haven't yet looked into it.
I'm still looking forward to contributions from you anyway!

**Comment by AndreRH** [comment]
I see patches for Qemu 7.2 in your repo, have you managed to make it work?

**Reply by miguelmazetto** [reply]
Not finished, i regressed it back to 5.2 after figuring it out that it wasn't qemu's fault, but I did start it. I'm not sure if the update is worth it now, after all 5.2 was released in 2020

**Reply by miguelmazetto** [reply]
Have you ever heard of the emulator [blink](https://github.com/jart/blink)?, i feel like it is better to use it instead of box86 or qemu. box86 has been focused in patching more and more applications instead of improving the general compatibility and speed, while blink is much more mature in the emulator correctness like qemu, but also is much faster than qemu. Unfortunally as of now there isnt a direct comparison between box86 and blink, but this should be considered instead of jumping to box86. box86 is often a hit or miss considering the amount of crashes in the issues.

**Reply by AndreRH** [reply]
> Not finished, i regressed it back to 5.2 after figuring it out that it wasn't qemu's fault, but I did start it. I'm not sure if the update is worth it now, after all 5.2 was released in 2020

I see, but please consider contributing your changes to qemu cpu "max", that seems useful!

**Reply by AndreRH** [reply]
> Have you ever heard of the emulator [blink](https://github.com/jart/blink)?, i feel like it is better to use it instead of box86 or qemu. box86 has been focused in patching more and more applications instead of improving the general compatibility and speed, while blink is much more mature in the emulator correctness like qemu, but also is much faster than qemu. Unfortunally as of now there isnt a direct comparison between box86 and blink, but this should be considered instead of jumping to box86. box86 is often a hit or miss considering the amount of crashes in the issues.

Haven't heard of it, but will definitely look into it, as FEX is quite a hard to integrate.

**Comment by Trass3r** [comment]
I guess the situation would be better now with the alternative emulators?

**Reply by miguelmazetto** [reply]
Last time i checked it was with wine 8.17, with all the alternative emulators as well.
Firefox still gives out the same error.
I only built with arm64/i386, so most of the programs tried to go x86_64 because the api said that 64 bits was supported, however the 64 bits came from the arm64. So... lots of problems. It is really hard to diagnose these things. I dont know how to debug wine, and even what i do debug it oftens is quite misterious. Alot of games i tried just hanged without doing anything.
It should be possible to have different results with armhf/i386, but the compile times are just brutal in my computer (15-30min).
If anyone wants to attempt it: [Termux Hangover](https://github.com/miguelmazetto/termux-packages-wine/tree/master/packages/hangover)
